salz-web/lib/game/conway.spec.js

```diff
@@ -0,0 +1,98 @@
+import Grid, { Cell } from './conway';
+
+describe('cell at coordinate (1, 2)', () => {
+  const cell = new Cell(1, 2);
+  test('has id -1 as the default owner', () => {
+    expect(cell.owner === -1).toBeTruthy();
+  });
+
+  test('can set owner to a different owner', () => {
+    cell.owner = 2;
+    expect(cell.owner === 2).toBeTruthy();
+  });
+});
+
+describe('game grid', () => {
+  const grid = new Grid(25);
+  const snapshot = {
+    turnid: 1,
+    cells: [
+      { x: 3, y: 3, playerid: 1 },
+      { x: 2, y: 3, playerid: 1 },
+      { x: 4, y: 3, playerid: 1 },
+    ],
+  };
+
+  test('can construct a grid of size 25', () => {
+    expect(grid.size === 25).toBeTruthy();
+    expect(grid.size !== 24).toBeTruthy();
+  });
+
+  test('starts with an empty collection of cells', () => {
+    expect(grid.cells.length === 0).toBeTruthy();
+  });
+
+  test('can mount a snapshot with 3 cells', () => {
+    grid.mountSnapshot(snapshot);
+    expect(grid.cells.length === 3).toBeTruthy();
+  });
+});
+
+describe('game rules', () => {
+  const grid = new Grid(50);
+  const snapshot = {
+    turnid: 1,
+    cells: [
+      { x: 3, y: 3, playerid: 1 },
+      { x: 2, y: 3, playerid: 1 },
+      { x: 4, y: 3, playerid: 1 },
+    ],
+  };
+  grid.mountSnapshot(snapshot);
+
+  const moves = [
+    {
+      turnid: 1,
+      moves: [
+        { x: 10, y: 10, playerid: 1 },
+        { x: 15, y: 15, playerid: 1 },
+        { x: 25, y: 25, playerid: 1 },
+      ],
+    },
+    {
+      turnid: 2,
+      moves: [
+        { x: 5, y: 2, playerid: 1 },
+        { x: 1, y: 4, playerid: 1 },
+        { x: 25, y: 25, playerid: 1 },
+      ],
+    },
+  ];
+
+  test('properly iterates to the first step', () => {
+    grid.next(moves[0].moves);
+    expect(grid.cells.length === 3).toBeTruthy();
+  });
+
+  test('properly iterates to the second step', () => {
+    grid.next(moves[1].moves);
+    expect(grid.cells.length === 3).toBeTruthy();
+    expect(grid.cells.filter(cell => cell.equals({ x: 4, y: 2 })).length !== 0).toBeTruthy();
+    expect(grid.cells.filter(cell => cell.equals({ x: 2, y: 4 })).length !== 0).toBeTruthy();
+  });
```

Comment:
I am not sure that this is right behavior.
If the snapshot is:
```
XXX
```
Then turn 1 with the commands should give:
```
 X
 X
 X
```
Now for turn 2, we apply the player commands [(5,2),(1,4),(25,25)] to the map that should contain [(2, 3, 1), (3, 3, 1), (4, 3, 1)]. Since none of the commands are within 1 of any active cell none of them apply. And then when we step the conways game of life rules we just get:
```
XXX
```
Let me know if I am missing something!

Reply:
> If the snapshot is:
> ```
> XXX
> ```
>
> Then turn 1 with the commands should give:
>```
> X
> X
> X
>```

Yep, that's checked on [line 72](https://github.com/Lugarun/salz/blob/cb708bed5babf750be684426eccded22b014320f/salz-web/lib/game/conway.spec.js#L72).
You'll see that it checks for 3 cells. I can change it to check for the coords of for the vertical pattern if that's better.
> Now for turn 2, we apply the player commands [(5,2),(1,4),(25,25)] to the map that should contain [(2, 3, 1), (3, 3, 1), (4, 3, 1)]. Since none of the commands are within 1 of any active cell none of them apply. And then when we step the conways game of life rules we just get:

Note the empty cells (4, 2) and (2, 4). Each of these cells have exactly 3 neighbours:
* Neighbours of (4, 2): (3, 2), **(5, 2)**, (3, 3)
* Neighbours of (2, 4): (3, 3), **(1, 4)**, (4, 3)

So both (4, 2) and (2, 4) should be born in the next round.

However, the cells (2, 3) and (4, 3) cannot be born due to over population (4 alive neighbours):
* Neighbours of (2, 3): (3, 2), (3, 3), **(1, 4)**, (3, 4)
* Neighbours of (4, 3): (3, 2), **(5, 2)**, (3, 3), (3, 4)

So the resulting pattern after turn 2 should be
```
  X
 X
X
```

Reply:
> Note the empty cells (4, 2) and (2, 4). Each of these cells have exactly 3 neighbours:
>
>    Neighbours of (4, 2): (3, 2), (5, 2), (3, 3)
>    Neighbours of (2, 4): (3, 3), (1, 4), (4, 3)

so the problem is that (5, 2) and (1, 4) shouldn't activate as they aren't directly beside a cell already owned by player 1, a player can only activate cells that are at most 1 away from a cell that he already has.

Reply:
Ah, okay. Can I have you write down all of the game rules somewhere in a file? Please put the file in the Telegram chat if you do.

In any case, clients don't have to check if a move is valid, right? The game engine will have ruled these out and so they will never appear in the database?

Reply:
The game engine doesn't do this yet